RISC-V: Update support for HiFive1 board

DieracDelta opened the pull request:
### Pull Request Overview

This is an attempt to merge MIT's branch of Tock into upstream for RISC-V, and then presumably master.
The major additions were:

- Corrected MMIO for PLIC (addresses seemed off when I compared to the memory map).
- Modified the PLIC's clear_all_pending(). Prior implementation wrote to read only register. Now claims and completes all interrupts.
- Integrated [an abstraction over writing to the CSR register space](https://github.com/DieracDelta/riscvregs). This allows us to write to the control/status registers in Rust and only need to specify the CSR location once--in its corresponding file. This makes the code both more readable and less bug prone. The library contains most of the m-mode and s-mode registers (u-mode ones shouldn't be difficult to implement), and registers necessary for the pmp. The source files are scattered throughout Tock, and used whever we write CSRs.
- Added in CLINT implementation; code source in arch/rv32i/clint.rs. Includes:
        - MMIO 
        - read/write for time/timecp
        - triggering software interrupts
        - Note that it seems like this memory region got replaced in the Arty-e21 board by just a timer, and you've abstracted over it with a machine timer. If we reduce the memory map for that machine timer (or distinguish it/include it as a parameter with different types), we could reuse the abstraction for both the clint and the machine-timer.
- An initial stab at a PMP driver using [this CSR abstraction](https://github.com/DieracDelta/riscvregs). Code source in arch/rv32i/pmp.rs and arch/rv32i/chip.rs. A couple of things to note:
        - I've implemented the bitfields in the library, so this should make things easier, as we can rely on this abstraction.
        - I don't have any way of testing this outside of qemu (only have e310 board), so would appreciate some help from people with a board.
        - Do we want to have the PMP implement the same things as the MPU? Presumably on most implementations, we can have it provide similar functionality. As it stands right now, the PMP is implementing the kernel::mpu::MPU trait. Does it make sense to refactor with the MPU renamed to something more general--perhaps a MemoryProtector trait--and have the PMP and the cortex-ms have the MPU and PMP implement that trait?
- Interrupt Handling. Added to the e310 board only (no way of testing the arty) a bit of pattern matching/more detailed debugging/error messages for exceptions and interrupts.
        - Claims and completes external interrupts.
        - Distinguishes each type of trap
        - Exceptions include debugging information.
- Slightly revised boards/hifive1/src/main.rs. Now, completely disables interrupts by clearing the mie bit in the mstatus csr, and the msoft, mtimer, and mext bits in the mie csr before modifying plic. If the goal is to supress all interrupts, we could easily include that functionality by just making the threshold greater than all the priorities.

### Testing Strategy

I was able to test on the Hifive1 board using the included Makefile. Oddly, was not able to get gdb to work with OpenOCD. Has anyone had success with this? I'd been running something along the lines of:

```
$PATH_TO_OPENOCD_RISCV_FORK -f $OPEN_OCD_CONFIG_FILE
$PATH_TO_RISCV_GDB $PATH_TO_HIFIVE1_ELF $PATH_TO_GDB_INIT_FILE
```
With a gdbinit file that looked like:
```
set remotetimeout 240
target extended-remote localhost:3333
monitor reset halt
monitor flash protect 0 64 off
load
```

It fails due to not having enough space to load into. Has anyone encountered this?


### TODO or Help Wanted

See prior section.


### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

bradjc commented on arch/rv32i/Cargo.toml:
```diff
@@ -7,3 +7,4 @@ edition = "2018"
 [dependencies]
 kernel = { path = "../../kernel" }
 tock_rt0 = { path = "../../libraries/tock-rt0" }
+riscvregs = { git = "https://github.com/DieracDelta/riscvregs.git", rev = "c03eb6c" }
```
Unfortunately currently Tock doesn't support (i.e. Tock doesn't allow) external dependencies because we cannot track `unsafe` use in dependencies currently. So, to use a library like this it would either need to be included in the `/library` folder in Tock, or included as a module in the rv32i crate.

DieracDelta replied:
Would it be reasonable to add this in as a git submodule under libraries?

alevy replied:
Ideally we'd avoid git submodules where possible (it's just a bigger hassle to develop with). Moving this into the repo would be great. You can technically do so without losing the git history with some git magic (`git filter-branch` i believe).

bradjc commented on chips/e310x/src/chip.rs:
```diff
@@ -136,16 +136,142 @@ impl kernel::Chip for E310x {
+                    None => (),
+                    Some(ext_interrupt_id) => {
+                        debug!("interrupt triggered {}\n", ext_interrupt_id);
+                        plic::complete(ext_interrupt_id);
```
This doesn't look quite right. The trap handler should only disable the interrupt, not mark it as finished. Or does this not work how I think?

DieracDelta replied:
A couple of things to note here. Spec is best described in section 10.7 [here](https://sifive.cdn.prismic.io/sifive%2F4d063bf8-3ae6-4db6-9843-ee9076ebadf7_fe310-g000.pdf). It's under-specified by the spec what happens if you claim the interrupt (read the claim/complete MMIO) multiple times. So at the very least I don't think we can rely on being able to handle the interrupt anytime after this. On the e310 board, from the testing I've done (for example, add in another claim before completing in this same spot), you get interrupt id 0 (no more pending interrupts). So we either need to claim, handle the interrupt, and complete the interrupt here, or have some global "in progress" state that we write to and then read from in service_pending_interrupts to complete. This seemed easier at the time. With this added knowledge, do you think it's worth changing? I can probably get a change out really quickly.

You're right, I should disable the interrupt here. 

Ideally, all interrupts should be suppressed after the reset_handler, so my intention was just to notify that the interrupt occurred (as it's not expected behavior), and then ignore it.

bradjc commented on arch/rv32i/src/plic.rs:
```diff
@@ -63,6 +67,12 @@ pub unsafe fn disable_all() {
     }
 }
 
+pub unsafe fn surpress_all() {
```
Add comment.

bradjc commented on arch/rv32i/src/lib.rs:
```diff
@@ -81,25 +84,24 @@ pub unsafe fn init_memory() {
-    :
-    : "volatile");
+/// The trap handler is called on exceptions and for interrupts.
+pub unsafe fn configure_trap_handler(mode: PermissionMode) {
```
It would be nice to keep the comment about how this is a generic implementation that specific boards may not want to use.

DieracDelta commented:
I added a couple of things to get c_hello to run (see most recent push):

- arch/rv32i/src/syscall.rs: the inline assembly snippet in switch_to_process was overwriting syscall_args. Specifically, when I compiled this for hifive1, the objdump showed something along the lines of:
```
2041aaa2 <_ecall> li    a5,0
2041aaa4 <_ecall+0x2> lw    t0,124(a6)
2041aaa8 <_ecall+0x6> addi    t0,t0,4
2041aaaa <_ecall+0x8> sw    t0,124(a6)
2041aaae <_ecall+0xc> lw    t0,36(a6)
2041aab2 <_ecall+0x10> sw    t0,0(a5)
2041aab6 <_ecall+0x14> lw    t0,40(a6)
2041aaba <_ecall+0x18> sw    t0,4(a5)
2041aabe <_ecall+0x1c> lw    t0,44(a6)
2041aac2 <_ecall+0x20> sw    t0,8(a5)
2041aac6 <_ecall+0x24> lw    t0,48(a6)
2041aaca <_ecall+0x28> sw    t0,12(a5)
2041aace <_ecall+0x2c> lw    t0,52(a6)
2041aad2 <_ecall+0x30> sw    t0,16(a5)
2041aad6 <_ecall+0x34> lw    a6,4(a6)
```
for the end of the inline assembly snippet in switch_to_process using the code as it currently stands in the riscv-pr branch. This is bad because ```li a5,0``` moves 0 in a5 but a5 should be holding syscall_args address. I'm not sure if this is a compiler bug, but it seems like one. The solution I'm proposing here is to simply save syscall_args and switch_reason on the stack before we mret and pop them off after. This fixed the issue when I compiled for the hifive1 board. To reproduce, clone the repo and build it for the hifive1 chip, objdump the elf, and grep for _ecall.

- boards/hifive1/layout.ld
I could not get tock to compile with reserved space for a userspace app. I think this comes down to the DTIM being too small on the hifive1 board. As a result, I artificially increased the DTIM on qemu and have been testing it on qemu. I've included instructions on how to do this. I also increased the space reserved for rom and prog. 

- arch/rv32i/src/lib.rs: When I compiled this for the hifive1 chip, the function had a prologue which pushed a register onto the stack. The #[naked] macro removed this prologue (thanks brad!). Additionally, in order to get the trap address written to mtvec I had to feed it in directly, not take the address of it.

- boards/hifive1/src/main.rs: I've included the same abstractions as in the arty board.

- chips/e310x/src/chip.rs: Replaced nullsyscall with the syscall from the arch/rv32i crate.

bradjc commented:
Hmm bors closed this...that was unexpected. @DieracDelta can you rebase this and re-open?
